core/sql/common/ComJSONFuncs.cpp

```diff
@@ -0,0 +1,793 @@
+    JsonSemAction *sem = (JsonSemAction *)malloc(sizeof(JsonSemAction));
+    GetState   *state = (GetState *)malloc(sizeof(GetState));
+    JsonReturnType ret;
+
+    memset(sem, 0, sizeof(JsonSemAction));
+    memset(state, 0, sizeof(GetState));
+
+    if(npath < 0)
+        return JSON_UNEXPECTED_ERROR;
+
+    state->lex = lex;
+    /* is it "_as_text" variant? */
+    state->normalize_results = normalize_results;
+    state->npath = npath;
+    state->path_names = tpath;
+    state->path_indexes = ipath;
+    state->pathok = (bool *)malloc(sizeof(bool) * npath);
+    state->array_cur_index = (int *)malloc(sizeof(int) * npath);
+
+    if (npath > 0)
+        state->pathok[0] = true;
+
+    sem->semstate = (void *) state;
+
+    /*
+     * Not all variants need all the semantic routines. Only set the ones that
+     * are actually needed for maximum efficiency.
+     */
+    sem->scalar = get_scalar;
+    if (npath == 0)
+    {
+        sem->object_start = get_object_start;
+        sem->object_end = get_object_end;
+        sem->array_start = get_array_start;
+        sem->array_end = get_array_end;
+    }
+    if (tpath != NULL)
+    {
+        sem->object_field_start = get_object_field_start;
+        sem->object_field_end = get_object_field_end;
+    }
+    if (ipath != NULL)
+    {
+        sem->array_start = get_array_start;
+        sem->array_element_start = get_array_element_start;
+        sem->array_element_end = get_array_element_end;
+    }
+
+    ret = pg_parse_json(lex, sem);
+    if (ret == JSON_OK)
+        *result = state->tresult;
+    else
+        *result = NULL;
+
+    if (lex != NULL)
+    {
+        if (lex->strval != NULL)
+        {
+            if (lex->strval->data != NULL)
+                free(lex->strval->data);
+            free(lex->strval);
+        }
+        free(lex);
+    }
+    if (sem != NULL)
+        free(sem);
+    if (state != NULL)
+        free(state);
+    return ret;
+}
+
+static JsonReturnType
+get_object_start(void *state)
+{
+    GetState   *_state = (GetState *) state;
+    int			lex_level = _state->lex->lex_level;
+
+    if (lex_level == 0 && _state->npath == 0)
+    {
+        /*
+         * Special case: we should match the entire object.  We only need this
+         * at outermost level because at nested levels the match will have
+         * been started by the outer field or array element callback.
+         */
+        _state->result_start = _state->lex->token_start;
+    }
+    return JSON_OK;
+}
+
+static JsonReturnType
+get_object_end(void *state)
+{
+    GetState   *_state = (GetState *) state;
+    int			lex_level = _state->lex->lex_level;
+
+    if (lex_level == 0 && _state->npath == 0)
+    {
+        /* Special case: return the entire object */
+        char	   *start = _state->result_start;
+        int			len = _state->lex->prev_token_terminator - start;
+
+        //_state->tresult = cstring_to_text_with_len(start, len);
+        _state->tresult = (char *)malloc(len + 1);
```

Comment:
It will be free in ex_function_json_object_field_text::eval().

    if (rltStr != NULL)
    {
        Lng32 rltLen = str_len(rltStr)+1;
        str_cpy_all(op_data[0], rltStr, rltLen);
        free(rltStr);

        // If result is a varchar, store the length of substring 
        // in the varlen indicator.
        if (getOperand(0)->getVCIndicatorLength() > 0)
            getOperand(0)->setVarLength(rltLen, op_data[-MAX_OPERANDS]);
    }